src/realm_dart_sync.cpp

```diff
@@ -33,20 +33,20 @@ RLM_API void realm_dart_http_request_callback(realm_userdata_t userdata, const r
     realm_http_request_t request_copy = request; // copy struct
 
     buf.url = request.url;
-    request_copy.url = buf.url.c_str();
     buf.body = std::string(request.body, request.body_size);
-    request_copy.body = buf.body.data();
-
     buf.headers_vector.reserve(request.num_headers);
     for (size_t i = 0; i < request.num_headers; i++) {
         auto [it, _] = buf.headers.emplace(request.headers[i].name, request.headers[i].value);
         buf.headers_vector.push_back({ it->first.c_str(), it->second.c_str() });
     }
-    request_copy.headers = buf.headers_vector.data();
 
     auto ud = reinterpret_cast<realm_dart_userdata_async_t>(userdata);
     ud->scheduler->invoke([ud, request_copy = std::move(request_copy), buf = std::move(buf), request_context]() {
-        (reinterpret_cast<realm_http_request_func_t>(ud->dart_callback)(ud->handle, request_copy, request_context));
+        realm_http_request_t request = std::move(request_copy);
```

Comment:
Why is this copy necessary?

Reply:
on line 45 we do need to copy it again since it's not modifiable since its an rvalue.

Reply:
In other words you get this error if you try to modify `request_copy.url`
`error C3490: 'url' cannot be modified because it is being accessed through a const object`

Reply:
Btw this ` realm_http_request_t request = std::move(request_copy);` is not a copy

Reply:
I'm not a c++ expert, but my understanding is that you cannot std::move a const object. So if the error you get is that you're accessing a const object, then `std::move`-ing it will just do a copy.

Reply:
std::move is not a copy its more of a const_cast

Reply:
Hmm.. to my understanding `std::move` on a const object will copy it. There was a discussion on stackoverflow about the same https://stackoverflow.com/questions/28595117/why-can-we-use-stdmove-on-a-const-object. Perhaps @RedBeard0531 can way in?

Reply:
I see @nirinchev already posted the exact same link, sorry 🤦

Reply:
The real answer here is that you want to add `mutable` to the lambda. Without it, all by-value captures, including `request_copy` are considered as `const` inside the body of the lambda. So `request` is silently making a copy, even though there is a `std::move`. You probably should also remove the const from the original `request` parameter, and the `request_copy` variable and just capture `request` as ~`request = std::move(request)`~~ simply `request` (see next paragraph). And if it the lambda is marked mutable, you can just modify it in place without making another copy. Since the callback will only be called once this is fine.

Oh. I just noticed that `realm_http_request_t` is the C API type. For C types (or any other "Trivially Copiable" type in C++) there is no difference between move and copy. So all of this discussion of move vs copy is moot, and you might as well just copy since it is syntactically lighter. I'm leaving what I've already written about `std::move` in the hope that it is informative.

> std::move is not a copy its more of a const_cast

To be pedantic, `const_cast` can only add or remove `const` and `volatile` qualifiers. `std::move` is specifically a `static_cast` that adds turns an lvalue into an rvalue. It does NOT remove the `const`. But it also doesn't copy (although the thing you pass the result to may make a copy of the referenced object).

> my understanding is that you cannot std::move a const object.

You _can_ `std::move` a const object, but you probably `shouldn't`. It produces a `const T&&` which is a bit of a weird thing. While a type _could_ have a special constructor just for that, in general it will select the copy constructor during overload resolution and make a copy. So the `move` isn't harmful (in that it doesn't change the behavior of the program, although it may make more template instantiations and potentially slow down compiles or bloat code size), but it is misleading. We could consider making our own `realm::move` that will cause a compiler error on const objects so that you know to either remove the `const` or remove the no-op `move`. The reason `std::move` allows passing `const` object is both for consistency, and because it is useful in templates where there are some cases where you want to move from non-const objects but obviously you can't from const ones, but you don't know what you will be passed and want to work with either. But that isn't something we do often.

Reply:
I knew there be dragons here.. Thanks you so much @RedBeard0531 🙏 

To be pragmatic, could you perhaps write a version of 
```c++
RLM_API void realm_dart_http_request_callback(realm_userdata_t userdata, const realm_http_request_t request, void* request_context)
```
as you think it should be written? So we have a hands on example of how theory manifest in practice?

Reply:
Based on my understanding, would something like this be fine:

```cpp
    ud->scheduler->invoke([ud, request, buf = std::move(buf), request_context]() {
        realm_http_request_t request_copy = request;
        request_copy.headers = buf.headers_vector.data();
        request_copy.url = buf.url.c_str();
        request_copy.body = buf.body.data();

        (reinterpret_cast<realm_http_request_func_t>(ud->dart_callback)(ud->handle, request_copy, request_context));
    });
```

Notably, we're capturing the original `request` and then creating the copy inside the callback.

Reply:
Also, if you add `mutable` to the lambda, you can get rid of the `realm_copy` variable.

BTW, if you are OK writing this part in C++, why not just implement the C++ `GenericNetworkTransport` API directly?

Reply:
Will fix

Reply:
Btw the C-API definition of the http request is this one 
```cpp
 /**
 * Callback function used by Core to make a HTTP request.
 *
 * Complete the request by calling realm_http_transport_complete_request(),
 * passing in the request_context pointer here and the received response.
 * Network request are expected to be asynchronous and can be completed on any thread.
 *
 * @param request The request to send.
 * @param request_context Internal state pointer of Core, needed by realm_http_transport_complete_request().
 */
typedef void (*realm_http_request_func_t)(realm_userdata_t userdata, const realm_http_request_t request,
                                          void* request_context);
```

Shouldn't that be a `const realm_http_request_t& request` instead. Why do they insist on copying the request object while invoking the function.

This is how they invoke it https://github.com/realm/realm-core/blob/master/src/realm/object-store/c_api/http.cpp#L60-L63. 

So why not pass the second argument by reference?

Reply:
@RedBeard0531 I did the fix. Could you please, check the PR again. Also this message above :)

Reply:
ok I spoke with Yavor about this. seems no point on doing that. there will be not much gain